File: go/300.500/438.FindAllAnagramsInaString.go
package main

import "fmt"

func findAnagrams(s string, p string) []int {
	ssize, psize := len(s), len(p)
	ret := []int{}
	if ssize < psize {
		return ret
	}
	scount := [26]int{}
	pcount := [26]int{}
	for i := 0; i < psize; i++ {
		scount[s[i]-'a']++
		pcount[p[i]-'a']++
	}
	if scount == pcount {
		ret = append(ret, 0)
	}
	n := ssize - psize
	for i := 0; i < n; i++ {
		scount[s[i]-'a']--
		scount[s[i+psize]-'a']++
		if scount == pcount {
			ret = append(ret, i+1)
		}
	}
	return ret
}

func findAnagrams1(s string, p string) []int {
	ssize, psize := len(s), len(p)
	ret := []int{}
	if ssize < psize {
		return ret
	}
	dp := [26]int{}
	for i, x := range p {
		dp[s[i]-'a']++
		dp[x-'a']--
	}
	differ := 0
	for _, x := range dp {
		if x != 0 {
			differ++
		}
	}
	if differ == 0 {
		ret = append(ret, 0)
	}
	n := ssize - psize
	for i := 0; i < n; i++ {
		x := s[i] - 'a'
		if dp[x] == 1 {
			differ--
		} else if dp[x] == 0 {
			differ++
		}
		dp[x]--
		x = s[i+psize] - 'a'
		if dp[x] == -1 {
			differ--
		} else if dp[x] == 0 {
			differ++
		}
		dp[x]++
		if differ == 0 {
			ret = append(ret, i+1)
		}
	}
	return ret
}

func main() {
	fmt.Println(findAnagrams1("af", "be"))
	fmt.Println(findAnagrams1("cbaebabacd", "abc"))
	fmt.Println(findAnagrams1("aaaaaaaaaa", "aaaaaaaaaaaaa"))
	fmt.Println(findAnagrams1("abab", "ab"))
	fmt.Println(findAnagrams1("aa", "bb"))
}

File: 300.500/438.FindAllAnagramsInaString.cpp
#include <iostream>
#include <string>
#include <vector>
#include <array>
using namespace std;

class FindAllAnagramsInaString
{
public:
    vector<int> findAnagrams(string s, string p) {
        //通过比较数组相等
        int psize = p.size();
        vector<int> ret;
        if (psize > s.size())
            return ret;
        array<int, 26> sCount{};
        array<int, 26> pCount{};
        for (int i = 0;i < psize;++i) {
            sCount[s[i] - 'a']++;
            pCount[p[i] - 'a']++;
        }
        if (sCount == pCount) {
            ret.emplace_back(0);
        }
        int n = s.size() - psize;
        for (int i = 0;i < n;++i) {
            sCount[s[i] - 'a']--;
            sCount[s[i + psize] - 'a']++;
            if (sCount == pCount)
                ret.emplace_back(i + 1);
        }
        return ret;
    }
public:
    vector<int> findAnagrams1(string s, string p) {
        //通过比较差异数为0
        int psize = p.size();
        vector<int> ret;
        if (psize > s.size())
            return ret;
        array<int, 26> count{};
        for (int i = 0;i < psize;++i) {
            ++count[s[i] - 'a'];
            --count[p[i] - 'a'];
        }
        int differ = 0;
        for (int x : count) {
            if (x != 0)
                differ++;
        }
        if (differ == 0)
            ret.emplace_back(0);
        int n = s.size() - psize;
        for (int i = 0;i < n;++i) {
            int x = s[i] - 'a';
            if (count[x] == 1) {
                --differ;
            }
            else if (count[x] == 0) {
                ++differ;
            }
            --count[x];
            x = s[i + psize] - 'a';
            if (count[x] == -1) {
                --differ;
            }
            else if (count[x] == 0) {
                ++differ;
            }
            ++count[x];
            if (differ == 0)
                ret.emplace_back(i + 1);
        }
        return ret;
    }
};

void Print(vector<int>& ret) {
    cout << "[";
    for (int x : ret) {
        cout << x << ", ";
    }
    cout << "]" << endl;
}

int main(int argc, char const* argv[])
{
    FindAllAnagramsInaString cls;
    auto& ret4 = cls.findAnagrams1("aaaaaaaaaa", "aaaaaaaaaaaaa");
    Print(ret4);
    auto& ret3 = cls.findAnagrams1("af", "be");
    Print(ret3);
    auto& ret = cls.findAnagrams1("cbaebabacd", "abc");
    Print(ret);
    auto& ret1 = cls.findAnagrams1("abab", "ab");
    Print(ret1);
    auto& ret2 = cls.findAnagrams1("aa", "bb");
    Print(ret2);
    return 0;
}
